File: server/datasets/java_source-code/Multiple-C2/organic_2.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #2
 * Space-optimized 1D DP array, iterating capacity in reverse.
 */

class Solution {
    public static int knapsack1d(int[] weights, int[] values, int capacity) {
        int[] dp = new int[capacity + 1];

        for (int i = 0; i < weights.length; i++) {
            int w_i = weights[i];
            int v_i = values[i];
            for (int cap = capacity; cap >= w_i; cap--) {
                dp[cap] = Math.max(dp[cap], dp[cap - w_i] + v_i);
            }
        }
        return dp[capacity];
    }

    public static void main(String[] args) {
        int[] itemsWeight = {1, 3, 4, 5};
        int[] itemsValue = {1, 4, 5, 7};
        int cap = 7;
        System.out.println(knapsack1d(itemsWeight, itemsValue, cap));
    }
}

File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_c2/organic_2.py
"""
0/1 Knapsack (DP) - Organic Submission #2
Space-optimized 1D DP array, iterating capacity in reverse.
"""


def knapsack_1d(weights, values, capacity):
    dp = [0] * (capacity + 1)

    for i in range(len(weights)):
        w_i, v_i = weights[i], values[i]
        for cap in range(capacity, w_i - 1, -1):
            dp[cap] = max(dp[cap], dp[cap - w_i] + v_i)

    return dp[capacity]


if __name__ == "__main__":
    items_weight = [1, 3, 4, 5]
    items_value = [1, 4, 5, 7]
    cap = 7
    print(knapsack_1d(items_weight, items_value, cap))
